python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -0,0 +1,184 @@
+        additional setup or teardown logic on each actor, so that the users of this
+        new trainer do not have to implement this logic. For example, a
+        ``TensorflowTrainer`` can be built on top of ``DataParallelTrainer``
+        that automatically handles setting the proper environment variables for
+        distributed Tensorflow on each actor.
+
+    For 1, you can set a predefined training loop in __init__
+
+    .. code-block:: python
+
+        from ray.ml.train.data_parallel_trainer import DataParallelTrainer
+
+        class MyDataParallelTrainer(DataParallelTrainer):
+            def __init__(self, *args, **kwargs):
+                predefined_train_loop_per_worker = lambda: 1
+                super().__init__(predefined_train_loop_per_worker, *args, **kwargs)
+
+
+    For 2, you can implement the ``ray.train.Backend`` and ``ray.train.BackendConfig``
+    interfaces.
+
+    .. code-block:: python
+
+        from dataclasses import dataclass
+        from ray.train.backend import Backend, BackendConfig
+
+        class MyBackend(Backend):
+            def on_start(self, worker_group, backend_config):
+                def set_env_var(env_var_value):
+                    import os
+                    os.environ["MY_ENV_VAR"] = env_var_value
+
+                worker_group.execute(set_env_var, backend_config.env_var)
+
+        @dataclass
+        class MyBackendConfig(BackendConfig):
+            env_var: str = "default_value"
+
+            def backend_cls(self):
+                return MyBackend
+
+        class MyTrainer(DataParallelTrainer):
+            def __init__(self, train_loop_per_worker, my_backend_config:
+                MyBackendConfig, **kwargs):
+
+                super().__init__(train_loop_per_worker, my_backend_config, **kwargs)
+
+    Args:
+        train_loop_per_worker: The training function to execute.
+            This can either take in no arguments or a ``config`` dict.
+        train_loop_config: Configurations to pass into
+            ``train_func`` if it accepts an argument.
+        backend_config: Used to specify which backend to setup on the
+            workers enable distributed communication, for example torch or
+            horovod. If no Backend should be set up, then set this to None.
+        scaling_config: Configuration for how to scale data parallel training.
+        run_config: Configuration for the execution of the training run.
+        datasets: Any Ray Datasets to use for training. Use
+            the key "train" to denote which dataset is the training
+            dataset. If a ``preprocessor`` is provided and has not already been fit,
+            it will be fit on the training dataset. All datasets will be transformed
+            by the ``preprocessor`` if one is provided.
+        preprocessor: A ray.ml.preprocessor.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
```

Comment:
Any concerns if we just rename this to `checkpoint` for simplicity?

Reply:
I believe this was previously discussed and the decision was to go with `resume_from_checkpoint`? In any case, I am going to keep as is for now. Please bring this up with the broader group for an API change.